linkerd/app/src/access_log_collector.rs

```diff
@@ -0,0 +1,192 @@
+use chrono::Utc;
+use futures::ready;
+use futures::task::{Context, Poll};
+use linkerd2_app_core::{drain, AccessLog};
+use linkerd2_error::Error;
+use pin_project::pin_project;
+use std::fmt::Debug;
+use std::future::Future;
+use std::pin::Pin;
+use std::time::Instant;
+use tokio::fs::{File, OpenOptions};
+use tokio::io::{AsyncWrite, BufWriter};
+use tokio::sync::mpsc;
+use tracing::{debug, warn};
+
+pub type Task = Pin<Box<dyn Future<Output = ()> + Send + 'static>>;
+
+pub type AccessLogSink = mpsc::Sender<AccessLog>;
+
+#[derive(Clone, Debug)]
+pub enum Config {
+    Disabled,
+    Enabled { path: String },
+}
+
+pub enum Collector {
+    Disabled,
+    Enabled { sink: AccessLogSink, task: Task },
+}
+
+impl Config {
+    const BUFFER_CAPACITY: usize = 100;
+
+    pub fn build(self, drain: drain::Watch) -> Result<Collector, Error> {
+        match self {
+            Config::Disabled => Ok(Collector::Disabled),
+            Config::Enabled { path } => {
+                let (sink, sink_rx) = mpsc::channel(Self::BUFFER_CAPACITY);
+
+                let task = Box::pin(async move {
+                    debug!("started access log collector");
+
+                    let f = OpenOptions::new()
+                        .create(true)
+                        .append(true)
+                        .open(path)
+                        .await
+                        .expect("Failed to open access log file");
+
+                    let writer = BufWriter::new(f);
+
+                    let sink = CollectorSink {
+                        writer,
+                        last_flush: Instant::now(),
+                        rx: sink_rx,
+                        state: State::Read,
+                    };
+
+                    if let Err(err) = drain.watch(sink, |sink| sink.shutdown()).await {
+                        warn!("failed to shutdown access log writer: {}", err);
+                    }
+                });
+
+                Ok(Collector::Enabled { sink, task })
+            }
+        }
+    }
+}
+
+impl Collector {
+    pub fn sink(&self) -> Option<AccessLogSink> {
+        match self {
+            Collector::Disabled => None,
+            Collector::Enabled { ref sink, .. } => Some(sink.clone()),
+        }
+    }
+}
+
+enum State {
+    Read,
+    Write(String, usize),
+    Flush,
+    Shutdown,
+}
+
+#[pin_project]
+struct CollectorSink {
+    #[pin]
+    writer: BufWriter<File>,
+    last_flush: Instant,
+    rx: mpsc::Receiver<AccessLog>,
+    state: State,
+}
+
+/// Consumes AccessLogs from receiver and writes them to the configured BufWriter
+impl CollectorSink {
+    /// A write FLUSH_TIMEOUT_SECS after the last flush will trigger a flush
+    const FLUSH_TIMEOUT_SECS: u64 = 1;
+
+    /// Transition future to shutting down. Future will complete with result
+    fn shutdown(&mut self) {
+        debug!("shutting down access log collector");
+        self.state = State::Shutdown
+    }
+}
+
+impl Future for CollectorSink {
```

Comment:
I think this could probably be expressed much more neatly as an `async fn`...unless the manual future implementation is necessary for the shutdown process? We could also just move the flush-on-shutdown behavior to a `Drop` guard that runs when the future is dropped...

Reply:
I'll have another play around, I couldn't work out how to make it work with the shutdown signal  :sweat_smile: